File: TIP Sample App/TweetImageFetchRequest.m
//
//  TweetImageFetchRequest.m
//  TwitterImagePipeline
//
//  Created on 2/11/17.
//  Copyright © 2017 Twitter. All rights reserved.
//

#import "AppDelegate.h"
#import "TweetImageFetchRequest.h"
#import "TwitterAPI.h"

#define kSMALL  @"small"
#define kMEDIUM @"medium"
#define kLARGE  @"large"

typedef struct {
    void * const name;
    CGFloat const dim;
} VariantInfo;

static VariantInfo const sVariantSizeMap[] = {
    { .name = kSMALL,   .dim = 680 },
    { .name = kMEDIUM,  .dim = 1200 },
    { .name = kLARGE,   .dim = 2048 },
};

static NSString *DetermineVariant(CGSize aspectRatio, const CGSize targetDimensions, UIViewContentMode targetContentMode);

@implementation TweetImageFetchRequest
{
    TweetImageInfo *_tweetImage;
    NSURL *_imageURL;
}

@synthesize targetDimensions = _targetDimensions;
@synthesize targetContentMode = _targetContentMode;

- (instancetype)initWithTweetImage:(TweetImageInfo *)tweetImage targetView:(UIView *)view
{
    if (self = [super init]) {
        _tweetImage = tweetImage;
        _targetContentMode = view.contentMode;
        _targetDimensions = TIPDimensionsFromView(view);
    }
    return self;
}

- (NSString *)imageIdentifier
{
    return _tweetImage.baseURLString;
}

- (NSURL *)imageURL
{
    if (!_imageURL) {
        if (self.forcePlaceholder) {
            _imageURL = [NSURL URLWithString:@"placeholder://placeholder.com/placeholder.jpg"];
        } else {
            NSString *URLString = nil;
            if ([_tweetImage.baseURLString hasPrefix:@"https://pbs.twimg.com/media/"]) {
                NSString *variantName = DetermineVariant(_tweetImage.originalDimensions, _targetDimensions, _targetContentMode);
                URLString = [NSString stringWithFormat:@"%@?format=%@&name=%@", _tweetImage.baseURLString, (APP_DELEGATE.searchWebP) ? @"webp" : _tweetImage.format, variantName];
            } else {
                URLString = [NSString stringWithFormat:@"%@.%@", _tweetImage.baseURLString, _tweetImage.format];
            }
            _imageURL = [NSURL URLWithString:URLString];
        }
    }
    return _imageURL;
}

- (TIPImageFetchOptions)options
{
    return (self.forcePlaceholder) ? TIPImageFetchTreatAsPlaceholder : TIPImageFetchNoOptions;
}

@end

static NSString *DetermineVariant(CGSize aspectRatio, const CGSize dimensions, UIViewContentMode contentMode)
{
    if (aspectRatio.height <= 0 || aspectRatio.width <= 0) {
        aspectRatio = CGSizeMake(1, 1);
    }

    const BOOL scaleToFit = (UIViewContentModeScaleAspectFit == contentMode);
    const CGSize scaledToTargetDimensions = TIPDimensionsScaledToTargetSizing(aspectRatio, dimensions, (scaleToFit ? UIViewContentModeScaleAspectFit : UIViewContentModeScaleAspectFill));

    NSString * selectedVariantName = nil;
    for (size_t i = 0; i < (sizeof(sVariantSizeMap) / sizeof(sVariantSizeMap[0])); i++) {
        const CGSize variantSize = CGSizeMake(sVariantSizeMap[i].dim, sVariantSizeMap[i].dim);
        const CGSize scaledToVariantDimensions = TIPDimensionsScaledToTargetSizing(aspectRatio, variantSize, UIViewContentModeScaleAspectFit);
        if (scaledToVariantDimensions.width >= scaledToTargetDimensions.width && scaledToVariantDimensions.height >= scaledToTargetDimensions.height) {
            selectedVariantName = (__bridge NSString *)sVariantSizeMap[i].name;
            break;
        }
    }

    if (!selectedVariantName) {
        selectedVariantName = kLARGE;
    }

    return selectedVariantName;
}

File: TIP Swift Sample App/TweetImageFetchRequest.swift
//
//  TweetImageFetchRequest.swift
//  TwitterImagePipeline
//
//  Created on 3/3/17.
//  Copyright © 2020 Twitter. All rights reserved.
//

//let kSMALL = "small"
//let kMEDIUM = "medium"
//let kLARGE = "large"
//
//struct VariantInfo {
//    var name: String
//    var dim: CGFloat
//
//    init(_ name: String, dimensions dim: CGFloat) {
//        self.name = name
//        self.dim = dim
//    }
//}
//
//let sVariantSizeMap: [VariantInfo] = [
//    VariantInfo.init(kSMALL, dimensions: 680),
//    VariantInfo.init(kMEDIUM, dimensions: 1200),
//    VariantInfo.init(kLARGE, dimensions: 2048)
//]

import TwitterImagePipeline

class TweetImageFetchRequest: NSObject, TIPImageFetchRequest {

    var forcePlaceholder: Bool

    private let tweetImageInfo: TweetImageInfo
    private var internalImageURL: URL?

    @objc var targetDimensions: CGSize

    @objc var targetContentMode: UIView.ContentMode

    @objc var imageIdentifier: String? {
        return self.tweetImageInfo.baseURLString
    }

    @objc var imageURL: URL {
        if nil == self.internalImageURL {
            if self.forcePlaceholder {
                self.internalImageURL = URL.init(string: "placeholder://placeholder.com/placeholder.jpg")
            } else {
                let URLString: String
                if self.tweetImageInfo.baseURLString.hasPrefix("https://pbs.twimg.com/media/") {
                    let variantName = TweetImageDetermineVariant(self.tweetImageInfo.originalDimensions, self.targetDimensions, self.targetContentMode)
                    let format = APP_DELEGATE().searchWebP ? "webp" : self.tweetImageInfo.format
                    URLString = "\(self.tweetImageInfo.baseURLString)?format=\(format)&name=\(variantName)"
                } else {
                    URLString = "\(self.tweetImageInfo.baseURLString).\(self.tweetImageInfo.format)"
                }
                self.internalImageURL = URL.init(string: URLString)
            }
        }

        return self.internalImageURL!
    }
    @objc var options: TIPImageFetchOptions {
        return self.forcePlaceholder ? [.treatAsPlaceholder] : []
    }

    init(tweetImage tweet: TweetImageInfo, targetView view: UIView?)
    {
        self.forcePlaceholder = false
        self.tweetImageInfo = tweet
        self.targetContentMode = view?.contentMode ?? .center
        self.targetDimensions = TIPDimensionsFromView(view)
    }
}
